### ISSUE TITLE
How can one define a new layer that have multiple inputs?

### ISSUE THREAD

**Issue by benjaminklein** [fchollet/keras#2364]
I want to define a new layer that have multiple inputs.
Each input has a different meaning and shape. How can I implement this layer using Keras?

**Comment by fchollet** [comment]
Use the functional API.

- Define a custom layer where the `call` method accepts a list of tensors (and may return a list of tensors, or just a single tensor).
- Use it like:

```python
c, d = layer([a, b])
```

Note that the `get_output_shape_for` method should also accept a list of shape tuples.

**Comment by benjaminklein** [comment]
Thank you. Can you give an example for a layer with a call method that accepts a list of tensors?

**Comment by carlthome** [comment]
@benjaminklein Look at the source. [Here for example](https://github.com/fchollet/keras/blob/5dad3786f6a34c15ce11ee862bab7a59ed713d68/keras/engine/topology.py#L1036).

**Comment by benjaminklein** [comment]
@carlthome Thank you.

**Comment by shirtzet** [comment]
I have the same problem.

I defined everything you described, but i get an error using it.
for example (very simple): 
lets say that the input is just one layer, and the "call" function returns "input". it works
say we accept multiple inputs, and we set the "call" function to return "input[0]". NOT WORKING
(get_output_shape_for is set accordingly)

Any ideas?

**Comment by THUfl12** [comment]
@benjaminklein, if your function doesn't contain any parameter to be learned from data. Here is an example, just put all the inputs into a list.

def concat_test(input):
    a = input[0]
    b = input[1]
    return K.concatenate([a, b], axis=1)

var_1 = Input(shape=(1,))
var_2 = Input(shape=(2,))

var = Lambda(concat_test, name='concat_test')([var_1, var_2])

But if your function has some parameters or weights needed to be trained, you'd better write your own layer. Since I am not very familiar with how to writing own layer, so I just give you the advice. Hope it helpful.

**Comment by THUfl12** [comment]
# define a self layer which doesn't include trainable parameters
class MyLayer(Layer):

    # initialize the layer, and set an extra parameter axis. No need to include inputs parameter!
    def __init__(self, axis, **kwargs):
        self.axis = axis
        self.result = None
        super(MyLayer, self).__init__(**kwargs)

    # first use build function to define parameters, Creates the layer weights.
    # input_shape will automatic collect input shapes to build layer
    def build(self, input_shape):
        print(input_shape)
        super(MyLayer, self).build(input_shape)

    # This is where the layer's logic lives. In this example, I just concat two tensors.
    def call(self, inputs, **kwargs):
        a = inputs[0]
        b = inputs[1]
        self.result = K.concatenate([a, b], axis=self.axis)
        return self.result

    # return output shape
    def compute_output_shape(self, input_shape):
        return K.int_shape(self.result)


var_1 = Input(shape=(1,))
var_2 = Input(shape=(2,))

var_test = MyLayer(axis=1)([var_1, var_2])

model = Model(inputs=[var_1, var_2], outputs=[var_test])
model.compile(optimizer='adam', loss='mse')

print(model.summary())
